src/ngx_lua_resty_lmdb_module.c

```diff
@@ -363,6 +375,151 @@ ngx_lua_resty_lmdb_verify_file_status(ngx_cycle_t *cycle,
 }
 
 
+static ngx_int_t
+ngx_lua_resty_lmdb_validate(ngx_cycle_t *cycle,
+                            ngx_lua_resty_lmdb_conf_t *lcf)
+{
+    int                        rc;
+    MDB_dbi                    dbi;
+    MDB_val                    key;
+    MDB_val                    value;
+    MDB_txn                   *txn = NULL;
+
+    ngx_str_t                  validation_key =
+                                    ngx_string(NGX_LUA_RESTY_LMDB_VALIDATION_KEY);
+
+    /* check tag value in lmdb */
+
+    if (lcf->validation_tag.data == NULL) {
+        return NGX_OK;
+    }
+
+    ngx_log_debug1(NGX_LOG_DEBUG_CORE, cycle->log, 0,
+                   "LMDB validation enabled, using validation tag: \"%V\"",
+                   &lcf->validation_tag);
+
+    ngx_lua_resty_lmdb_assert(lcf->validation_tag.data);
+
+    rc = mdb_txn_begin(lcf->env, NULL, 0, &txn);
+    if (rc != 0) {
+        ngx_log_error(NGX_LOG_CRIT, cycle->log, 0,
+                      "unable to open LMDB transaction: %s",
+                      mdb_strerror(rc));
+        return NGX_ERROR;
+    }
+
+    ngx_lua_resty_lmdb_assert(txn);
+
+    rc = mdb_dbi_open(txn, NGX_LUA_RESTY_LMDB_DEFAULT_DB, MDB_CREATE, &dbi);
+    if (rc != 0) {
+        ngx_log_error(NGX_LOG_ERR, cycle->log, 0,
+                      "unable to open LMDB database: %s",
+                      mdb_strerror(rc));
+        goto failed;
+    }
+
+    key.mv_size = validation_key.len;
+    key.mv_data = validation_key.data;
+
+    rc = mdb_get(txn, dbi, &key, &value);
+    if (rc == 0) {
+        /* key found, compare with validation_tag value */
+        if (lcf->validation_tag.len == value.mv_size &&
+            ngx_strncmp(lcf->validation_tag.data,
+                        value.mv_data, value.mv_size) == 0) {
+
+            mdb_txn_abort(txn);
+            return NGX_OK;
+        }
+
+        ngx_log_error(NGX_LOG_WARN, cycle->log, 0,
+                      "LMDB tag \"%*s\" did not match configured tag \"%V\"",
+                      value.mv_size, value.mv_data,
+                      &lcf->validation_tag);
+
+    } else if (rc == MDB_NOTFOUND) {
+        ngx_log_error(NGX_LOG_WARN, cycle->log, 0,
+                      "LMDB validation tag does not exist");
+
+    } else {
+        ngx_log_error(NGX_LOG_ERR, cycle->log, 0,
+                      "unable to get LMDB validation_tag : %s",
```

Comment:
Please make sure error messages are consistent. They should all be "LMDB validation tag", not "LMDB tag" and "LMDB validation_tag".

Reply:
done